### PR TITLE
Refactor internal error handling to use exceptions

### PR DESCRIPTION
**Pull request by MikeGitb**
Extracted the exception based error handling from #48 and put it in this separate PR.

Disclaimer: I didn't run any tests on this other than that it compiles (I don't have a hardware device)

### CODE REVIEW

File: libsweep/include/Error.hpp [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#ifndef SWEEP_ERROR_2EADE195E243_H
+#define SWEEP_ERROR_2EADE195E243_H
+
+#include <stdexcept>
+
+namespace sweep {
+    struct ErrorBase : std::runtime_error {
```
Comment (daniel-j-h) [inline libsweep/include/Error.hpp]: Can you please lowercase the file to `error.h` as well as `error` here: it's already in the `sweep` namespace, so it's the generic `sweep::error` type.

File: libsweep/src/protocol.cc [hunk 2]
```diff
@@ -2,10 +2,15 @@
 #include <stdlib.h>
 
 #include "protocol.h"
+#include "../include/Error.hpp"
 
 namespace sweep {
 namespace protocol {
 
+struct Error : ::sweep::ErrorBase {
```
Comment (daniel-j-h) [inline libsweep/src/protocol.cc]: Same here

File: libsweep/src/protocol.cc [hunk 3]
```diff
@@ -2,10 +2,15 @@
 #include <stdlib.h>
 
 #include "protocol.h"
+#include "../include/Error.hpp"
```
Comment (daniel-j-h) [inline libsweep/src/protocol.cc]: Please no relative include paths, the include dir should be on the search path already (check CMakeLists.txt)

File: libsweep/src/sweep.cc [hunk 4]
```diff
@@ -1,15 +1,18 @@
 #include "sweep.h"
 #include "protocol.h"
 #include "serial.h"
+#include "../include/Error.hpp"
```
Comment (daniel-j-h) [inline libsweep/src/sweep.cc]: Same here

File: libsweep/src/sweep.cc [hunk 5]
```diff
@@ -19,25 +22,17 @@ typedef struct sweep_device {
 
 #define SWEEP_MAX_SAMPLES 4096
 
-typedef struct sweep_scan {
+struct sweep_scan {
```
Comment (daniel-j-h) [inline libsweep/src/sweep.cc]: Hm if you want to remove typedefs feel free to do it for all structs

File: libsweep/src/sweep.cc [hunk 6]
```diff
@@ -46,149 +41,102 @@ void sweep_error_destruct(sweep_error_s error) {
...
-
-sweep_device_s sweep_device_construct(const char* port, int32_t bitrate, sweep_error_s* error) {
+// internal implementations of functions - can throw exceptions
+namespace {
+
+struct Error : ::sweep::ErrorBase {
+  using ::sweep::ErrorBase::ErrorBase;
+};
+
+// forward declarations for internal definitions
```
Comment (daniel-j-h) [inline libsweep/src/sweep.cc]: Why? We include the `sweep.h` header anyway, no?

File: libsweep/src/sweep.cc [hunk 7]
```diff
@@ -46,149 +41,102 @@ void sweep_error_destruct(sweep_error_s error) {
...
-  return sweep_device_construct("/dev/ttyUSB0", 115200, error);
-}
-
-sweep_device_s sweep_device_construct(const char* port, int32_t bitrate, sweep_error_s* error) {
+// internal implementations of functions - can throw exceptions
+namespace {
+
+struct Error : ::sweep::ErrorBase {
+  using ::sweep::ErrorBase::ErrorBase;
+};
```
Comment (daniel-j-h) [inline libsweep/src/sweep.cc]: Shouldn't we use the base from error.h here?

File: libsweep/src/sweep.cc [hunk 8]
```diff
@@ -46,149 +41,102 @@ void sweep_error_destruct(sweep_error_s error) {
...
-    sweep::serial::error_destruct(serialerror);
-    return nullptr;
-  }
 
-  auto out = new sweep_device{serial, /*is_scanning=*/true};
+  sweep::serial::device_s serial = sweep::serial::device_construct(port, bitrate);
 
-  sweep_error_s stoperror = nullptr;
-  sweep_device_stop_scanning(out, &stoperror);
+  auto out = std::unique_ptr<sweep_device>(new sweep_device{serial, /*is_scanning=*/true});
```
Comment (daniel-j-h) [inline libsweep/src/sweep.cc]: Why the unique ptr here? For cleaning up in case of exceptions? This is a bit ugly

File: libsweep/src/sweep.cc [hunk 9]
```diff
@@ -46,149 +41,102 @@ void sweep_error_destruct(sweep_error_s error) {
...
-    return nullptr;
-  }
+  sweep_device_stop_scanning(out.get());
 
-  return out;
+  return out.release();
 }
 
-void sweep_device_destruct(sweep_device_s device) {
+void sweep_device_destruct_impl(sweep_device_s device) {
```
Comment (daniel-j-h) [inline libsweep/src/sweep.cc]: Why the change to _impl here?

File: libsweep/src/sweep.cc [hunk 10]
```diff
@@ -212,7 +160,7 @@ sweep_scan_s sweep_device_get_scan(sweep_device_s device, sweep_error_s* error)
     }
   }
 
-  auto out = new sweep_scan;
+  auto out = std::unique_ptr<sweep_scan>(new sweep_scan);
```
Comment (daniel-j-h) [inline libsweep/src/sweep.cc]: Same here

File: libsweep/src/sweep.cc [hunk 11]
```diff
@@ -395,38 +262,132 @@ void sweep_device_set_sample_rate(sweep_device_s device, int32_t hz, sweep_error
...
-  sweep::protocol::response_param_s response;
-  sweep::protocol::read_response_param(device->serial, sweep::protocol::SAMPLE_RATE_ADJUST, &response, &protocolerror);
+  sweep::protocol::write_command(device->serial, sweep::protocol::RESET_DEVICE);
+}
 
-  if (protocolerror) {
-    *error = sweep_error_construct("unable to receive sample rate command response");
-    sweep::protocol::error_destruct(protocolerror);
-    return;
+template <class F> auto translateException(F&& f, sweep_error_s* error) -> decltype(std::forward<F>(f)()) {
```
Comment (daniel-j-h) [inline libsweep/src/sweep.cc]: The following should probably be in the error.h header

File: libsweep/src/sweep.cc [hunk 12]
```diff
@@ -395,38 +262,132 @@ void sweep_device_set_sample_rate(sweep_device_s device, int32_t hz, sweep_error
...
+  try {
+    return std::forward<F>(f)();
+  } catch (const sweep::ErrorBase& e) {
+    *error = new sweep_error{e.what()};
+  } catch (const std::bad_alloc&) {
```
Comment (daniel-j-h) [inline libsweep/src/sweep.cc]: Why not catching `std::exception&` for which there is a `.what()`

File: libsweep/src/sweep.cc [hunk 13]
```diff
@@ -395,38 +262,132 @@ void sweep_device_set_sample_rate(sweep_device_s device, int32_t hz, sweep_error
...
+    *error = new sweep_error{"Could not allocate enough memory"};
+  } catch (...) {
+    *error = new sweep_error{"Unknown exception"};
+  }
+  return {};
+}
+
+template <class F> void translateExceptionV(F&& f, sweep_error_s* error) {
```
Comment (daniel-j-h) [inline libsweep/src/sweep.cc]: Would be great if we can overload this function instead

File: libsweep/src/sweep.cc [hunk 14]
```diff
@@ -395,38 +262,132 @@ void sweep_device_set_sample_rate(sweep_device_s device, int32_t hz, sweep_error
...
+    *error = new sweep_error{e.what()};
+  } catch (const std::bad_alloc&) {
+    *error = new sweep_error{"Could not allocate enough memory"};
+  } catch (...) {
+    *error = new sweep_error{"Unknown exception"};
   }
 }
+} // local namespace
+
+/* ###################### implementation of public API ################################# */
```
Comment (daniel-j-h) [inline libsweep/src/sweep.cc]: Ah now I understand - you split off the C API and the internal functions. Hm I see no benefit in doing so, what's the reason behind this?

File: libsweep/src/unix/serial.cc [hunk 15]
```diff
@@ -1,6 +1,7 @@
 #define _POSIX_C_SOURCE 200809L
 
 #include "serial.h"
+#include "../../include/Error.hpp"
```
Comment (daniel-j-h) [inline libsweep/src/unix/serial.cc]: Same here

File: libsweep/include/Error.hpp (hunk 1 above)
Comment (MikeGitb) [inline libsweep/include/Error.hpp]: OK on the lower case, but removing `Base` will let it collide with the error defined in sweep.cc.

File: libsweep/src/sweep.cc (hunk 14 above)
Comment (MikeGitb) [inline libsweep/src/sweep.cc]: Separation of concerns: The private functions are concerned with implementing the functionality the implementation of the public functions then only needs to take care about eror translation and any other adaptions to the ABI that might be necessary in the future. Also `translateException` needs some function / function object to be passed to it anyway, so the only alternative I see would be to wrapp the internal functionality into a lambda which doesn't really make the code more readable.

File: libsweep/src/sweep.cc (hunk 13 above)
Comment (MikeGitb) [inline libsweep/src/sweep.cc]: As I said in the other thread: If you see a simple way to do this be my guest. also I really don't see what the advantage would be.

File: libsweep/src/sweep.cc (hunk 12 above)
Comment (MikeGitb) [inline libsweep/src/sweep.cc]: Overlooked, will fix

File: libsweep/src/sweep.cc (hunk 11 above)
Comment (MikeGitb) [inline libsweep/src/sweep.cc]: As it is translating an exception into and ABI error, which is defined in `sweep.h` and no other part of the code has to care about this I don't see why this should be moved there.

File: libsweep/src/sweep.cc (hunk 9 above)
Comment (MikeGitb) [inline libsweep/src/sweep.cc]: To avoid naming collision (the API I version doesn't hav an error parameter to distinguish between the two). See comment below about why I have two in the first place

File: libsweep/src/sweep.cc (hunk 8 above)
Comment (MikeGitb) [inline libsweep/src/sweep.cc]: Precisely. `sweep_device_stop_scanning` can throw and besides it is my policy to not have a naked `new` anywhere in my code if I can at all help it. 

What would be your suggestion (in c++14 you can use make_unique)?

File: libsweep/src/sweep.cc (hunk 6 above)
Comment (MikeGitb) [inline libsweep/src/sweep.cc]: See comment at the end

File: libsweep/src/sweep.cc (hunk 7 above)
Comment (MikeGitb) [inline libsweep/src/sweep.cc]: That would also be a possibility. 
In general I wanted to distinguish between exceptions thrown in the library in general and exceptions that are thrown on this particular level (e.g. as opposed to the IO or protocol level). But currently there is admittedly nothing that makes use of that distinction. We could even use the same exception type everywhere in the code.

File: libsweep/src/sweep.cc (hunk 5 above)
Comment (MikeGitb) [inline libsweep/src/sweep.cc]: Must have slipped through, will fix. 
I originally had all the typedefs removed throughout the library but wanted to keep this PR clean of it (there is already going on quite a lot anyway), so I reverted those changes - except that I apparently overlooked this one.

File: libsweep/src/protocol.cc (hunk 3 above)
Comment (MikeGitb) [inline libsweep/src/protocol.cc]: Can do.
Although I'm really no fan of unqualified includes for files outside the same directory - in particular for files with very generic names.

**Comment by MikeGitb** [comment]
@daniel-j-h: Ok, I
- made `error.hpp` and `error` lower case
- changed to unqualified includes
- changed the error translation a bit

I did not incorporate the rest of your feedback and am waiting for your responses to my comments instead. I'm not sure, why git marks them all as outdated

File: libsweep/src/sweep.cc (hunk 5 above)
Comment (MikeGitb) [inline libsweep/src/sweep.cc]: Actually I think I'll just leave it for now and provide another PR once / if this one gets merged

**Comment by MikeGitb** [comment]
Rebased onto master

**Comment by MikeGitb** [comment]
Rebased on master and rewrote error handling in win/serial.cc to use exceptions

**Comment by MikeGitb** [comment]
So ist there any chance this will get merged in the near future?

**Comment by daniel-j-h** [comment]
Yes, please excuse my inactivity I'm out until late next week. Maybe
@dcyoung can jump in here - otherwise you will have to wait a bit. I
want to get us to switch to exceptions and references for sure.

**Comment by MikeGitb** [comment]
No rush. Main question is, if you are OK with the general structure of this PR as it is now.

**Comment by MikeGitb** [comment]
Closing this for now - I think after the last round of changes a merge is no longer possible. I won't delete the branch though, as parts might be salvageable for future PRs
